### PR TITLE
Add HTTP-verb methods on ajax service

### PR DESCRIPTION
**Pull request by dasnixon**
* `post(url, options)` - sets options.type to POST
* `del(url, options)` - sets options.type to DELETE
* `put(url, options)` - sets options.type to PUT

Added documentation and a few comments in code

### CODE REVIEW

**Comment by MattNguyen** [comment]
+1

File: tests/unit/services/ajax-test.js [hunk 1]
```diff
@@ -134,6 +134,49 @@ test("request() promise label is correct", function(assert) {
   assert.equal(postPromise._label, 'ember-ajax: POST to /posts');
 });
 
+test("post() promise label is correct", function(assert) {
+  service = Service.create();
+  const url = '/posts';
```
Comment (stefanpenner) [inline tests/unit/services/ajax-test.js]: these all seem like `let`.

`const` should be reserved for things that absolutely must stay constant, not `let` that  happen to remain constant.
Individuals should feel compelled to NOT change `const` -> `let`, over utilizing it degrades the value of this messaging.

File: tests/unit/services/ajax-test.js [hunk 2]
```diff
@@ -134,6 +134,49 @@ test("request() promise label is correct", function(assert) {
...
+  const options = {
+    data: {
+      post: { title: 'Title', description: 'Some description.' }
+    }
+  };
+  const serverResponse = [200, { "Content-Type": "application/json" }, JSON.stringify(options.data)];
+
+  server.post(url, () => serverResponse);
+
+  const postPromise = service.post(url, options);
```
Comment (stefanpenner) [inline tests/unit/services/ajax-test.js]: we should actually validate the responses.

```js
return service.post(url, options).then( more assetions)
```

File: tests/unit/services/ajax-test.js [hunk 3]
```diff
@@ -134,6 +134,49 @@ test("request() promise label is correct", function(assert) {
...
+  const options = {
+    data: {
+      post: { title: 'Title', description: 'Some description.' }
+    }
+  };
+  const serverResponse = [200, { "Content-Type": "application/json" }, JSON.stringify(options.data)];
+
+  server.put(url, () => serverResponse);
+
+  const putPromise = service.put(url, options);
```
Comment (stefanpenner) [inline tests/unit/services/ajax-test.js]: ^

File: tests/unit/services/ajax-test.js [hunk 4]
```diff
@@ -134,6 +134,49 @@ test("request() promise label is correct", function(assert) {
...
+});
+
+test("del() promise label is correct", function(assert) {
+  service = Service.create();
+  const url = '/posts/1';
+  const serverResponse = [200, { "Content-Type": "application/json" }, JSON.stringify({})];
+
+  server.delete(url, () => serverResponse);
+
+  const delPromise = service.del(url);
```
Comment (stefanpenner) [inline tests/unit/services/ajax-test.js]: ^

File: tests/unit/services/ajax-test.js (hunk 1 above)
Comment (elwayman02) [inline tests/unit/services/ajax-test.js]: As I understand it, `const` means "constant reference", not "constant value", @stefanpenner...is this wrong?

Comment (elwayman02) [inline tests/unit/services/ajax-test.js]: More specifically, I'm based that off the MDN description of `const`: 
> The const declaration creates a read-only reference to a value. It does not mean the value it holds is immutable, solely that the variable identifier can not be reassigned.

It seems like you are saying that we should treat `const` as if it is immutable, which is the opposite of what its documentation claims.

File: addon/services/ajax.js [hunk 5]
```diff
@@ -133,6 +133,29 @@ export default Ember.Service.extend({
...
+
+  // calls `request()` but forces `options.type` to `DELETE`
+  del(url, options) {
+    return this.request(url, this._addTypeToOptionsFor(options, 'DELETE'));
+  },
+
+  // forcibly manipulates the options hash to include the HTTP method on the type key
+  _addTypeToOptionsFor(options, method) {
+    options = options || {};
+    delete options.type;
```
Comment (elwayman02) [inline addon/services/ajax.js]: Why do you need to delete type right before you reassign it? Couldn't you just have line 155 on its own?

Comment (stefanpenner) [inline addon/services/ajax.js]: ya no point.

File: tests/unit/services/ajax-test.js (hunk 1 above)
Comment (stefanpenner) [inline tests/unit/services/ajax-test.js]: No, I specifically made no reference to immutability. Local vars that "happen" to not change, are not `const`, they are let. `const` is meant to convey, please do to rebind this given variable.

Comment (elwayman02) [inline tests/unit/services/ajax-test.js]: Can you give an example of when you would want to use `const`, then?

Comment (stefanpenner) [inline tests/unit/services/ajax-test.js]:
```

const USER_URL = 'http:...../'

export default Ember.Object.extend({

  actions: {
     save(user) {
        post(USER_URL + '/' + user.get('id'));
     }
  }
});
```

Comment (elwayman02) [inline tests/unit/services/ajax-test.js]: That's an example of something that, in any other language, would be considered immutable. Despite you not making a specific reference to immutability, you are still asserting that `const` should only be used in this type of scenario. Thus, I am confused why the ES6 spec bothered to introduce a keyword that is apparently *supposed* to convey this concept while specifically calling for a different implementation altogether. That seems relatively useless, does it not?

Comment (stefanpenner) [inline tests/unit/services/ajax-test.js]: let/const/var describing the variable binding, not the content.

const is immutable binding, let is block scoped binding, var is yolo binding.

Comment (stefanpenner) [inline tests/unit/services/ajax-test.js]: > That's an example of something that, in any other language, would be considered immutable.

I don't believe thats actually the case.

Comment (elwayman02) [inline tests/unit/services/ajax-test.js]: Ok. I don't know that I'm entirely convinced of the usefulness of that distinction, but it's not an unreasonable explanation. Thanks for clarifying, @stefanpenner!

Comment (stefanpenner) [inline tests/unit/services/ajax-test.js]: It's subtle, but has value.

**Comment by dasnixon** [comment]
@elwayman02 @stefanpenner updated, `let`, me know your thoughts (get the pun). I think I used `let`/`const` more correctly?

File: tests/unit/services/ajax-test.js [hunk 6]
```diff
@@ -127,13 +127,79 @@ test("request() promise label is correct", function(assert) {
...
+  };
+
+  const serverResponse = [200, { "Content-Type": "application/json" }, JSON.stringify(options.data)];
+
+  server.post(url, () => serverResponse);
+
+  const postPromise = service.post(url, options);
+  assert.equal(postPromise._label, 'ember-ajax: POST to /posts');
+
+  postPromise.then(function (response) {
```
Comment (stefanpenner) [inline tests/unit/services/ajax-test.js]: these need to return, to be async tests.

File: tests/unit/services/ajax-test.js [hunk 7]
```diff
@@ -127,13 +127,79 @@ test("request() promise label is correct", function(assert) {
...
+  };
+
+  const serverResponse = [200, { "Content-Type": "application/json" }, JSON.stringify(options.data)];
+
+  server.put(url, () => serverResponse);
+
+  const putPromise = service.put(url, options);
+  assert.equal(putPromise._label, 'ember-ajax: PUT to /posts/1');
+
+  putPromise.then(function (response) {
```
Comment (stefanpenner) [inline tests/unit/services/ajax-test.js]: ^

File: tests/unit/services/ajax-test.js [hunk 8]
```diff
@@ -127,13 +127,79 @@ test("request() promise label is correct", function(assert) {
...
+  let url = '/posts/1';
+
+  const serverResponse = [200, { "Content-Type": "application/json" }, JSON.stringify({})];
+
+  server.delete(url, () => serverResponse);
+
+  const delPromise = service.del(url);
+  assert.equal(delPromise._label, 'ember-ajax: DELETE to /posts/1');
+
+  delPromise.then(function (response) {
```
Comment (stefanpenner) [inline tests/unit/services/ajax-test.js]: ^

**Comment by dasnixon** [comment]
@stefanpenner updated and used `deepEqual` to compare the `response` from the promise to the `options` object.

**Comment by stefanpenner** [comment]
@dasnixon thank you kindly, LVGTM

@embersherpa r?

**Comment by taras** [comment]
@stefanpenner I don't have an opinion on this, so if you think it's good then I'll merge and release it.

**Comment by taras** [comment]
@dasnixon thank you for your contribution and @stefanpenner thank you for the code review :)

**Comment by taras** [comment]
Just published 0.7.0 with this PR. Thank you :)

**Comment by dasnixon** [comment]
Thanks I appreciate it!

**Pull request merged**